src/reanimated2/hook/useScrollViewOffset.ts

```diff
@@ -89,15 +90,30 @@ function useScrollViewOffsetNative(
 
   useEffect(() => {
     // We need to make sure that listener for old animatedRef value is removed
-    if (scrollRef.current !== null) {
-      eventHandler.workletEventHandler.unregisterFromEvents();
+    if (scrollRef.current !== null && scrollRefTag.current !== null) {
+      eventHandler.workletEventHandler.unregisterFromEvents(
+        scrollRefTag.current
+      );
     }
+
+    // Store the ref and viewTag for future cleanup
     scrollRef.current = animatedRef.current;
+    scrollRefTag.current = animatedRef.getTag();
+
+    if (scrollRefTag === null) {
+      console.warn(
+        '[Reanimated] ScrollViewOffset failed to resolve the view tag from animated ref. Did you forget to attach the ref to a component?'
+      );
+    } else {
+      eventHandler.workletEventHandler.registerForEvents(scrollRefTag.current);
+    }
 
-    const viewTag = animatedRef.getTag();
-    eventHandler.workletEventHandler.registerForEvents(viewTag);
     return () => {
-      eventHandler.workletEventHandler.unregisterFromEvents();
+      if (scrollRefTag.current !== null) {
```

Comment:
Note generally reading a mutable reference in effect cleanup is risky because it could've changed by the time you got to cleanup (e.g. it could've been ... cleaned up). Or replaced by a newer reference. So it's a good idea to do

```js
useEffect(() => {
  const instance = someRef.current
  doStuff(instance)
  return () => {
    undoStuff(instance)
  })
})
```

Reply:
Yes, thanks for pointing out - luckily newer versions have already made it this way 😄